File: day 13 -25/day_23/day_23.py
import time

s = time.time()

inpt = '389125467'
cups = [int(cup) for cup in inpt] + [i for i in range(10, 10 ** 6 + 1)]

MAX = max(cups)
MIN = min(cups)

# Part 1
current_cup_id = 0
current_cup = cups[current_cup_id]
for i in range(1):
    #print('Cups: {}'.format(['({})'.format(cup) if cup == current_cup else cup for cup in cups]))
    
    three_cups = [cups[(current_cup_id + i) % len(cups)] for i in range(1, 4)]
    #print('Three cups: {}'.format(three_cups))
    
    destination_cup = current_cup
    while True:
        destination_cup -= 1
        if destination_cup < MIN:
            destination_cup = MAX 
        if destination_cup not in three_cups:
            #print('Destination cup: {}'.format(destination_cup))
            for i in range(3): del cups[(cups.index(current_cup) + 1) % len(cups)]
            insert_id = cups.index(destination_cup) + 1
            break
    cups = cups[:insert_id] + three_cups + cups[insert_id:]
        
    current_cup_id = (cups.index(current_cup) + 1) % len(cups)
    current_cup = cups[current_cup_id]
        
print(time.time() - s)

File: day 13 -25/day_23/day_23.c
#include <stdio.h>
#define SIZE 9
#define LOOP_SIZE 10

void printCups(int *cups) {
    for (int i = 0; i < SIZE; i++ ) { printf("%d ", cups[i]); }
    printf("\n");
}

int main() {
    int cups[SIZE] = {3, 8, 9, 1, 2, 5, 4, 6, 7};
    //printCups(cups);

    for (int i = 0; i < LOOP_SIZE; i++) {
        int currentCup = cups[0];
        int nextCurrentCup = cups[4];
        int threeCups[]    = {cups[1], cups[2], cups[3]};
        
        int target = currentCup - 1;
        int j = 1;
        while (j < 4) {
            if (target < 1) { target = SIZE; }
            if (cups[j] == target) { 
                target--; 
                j = 1;
            } else {
                j += 1;
            }
        }
        int h = 4;
        int delay = 4;
        while (h < SIZE) {
            cups[h - delay] = cups[h];
            if (cups[h] == target) {
                cups[h - 3] = threeCups[0];
                cups[h - 2] = threeCups[1];
                cups[h - 1] = threeCups[2];
                delay = 1;
            }
            h += 1;
        }
        cups[SIZE - 1] = currentCup;
        printCups(cups);
    }
    return 0;
}
